src/Admin/Product/ChannelVisibilityBlock.php

```diff
@@ -0,0 +1,134 @@
+<?php
+declare( strict_types=1 );
+
+namespace Automattic\WooCommerce\GoogleListingsAndAds\Admin\Product;
+
+use Automattic\WooCommerce\GoogleListingsAndAds\Product\ProductHelper;
+use Automattic\WooCommerce\GoogleListingsAndAds\Product\ProductSyncer;
+use Automattic\WooCommerce\GoogleListingsAndAds\Value\ChannelVisibility;
+use Automattic\WooCommerce\GoogleListingsAndAds\Value\SyncStatus;
+use WP_REST_Request as Request;
+use WP_REST_Response as Response;
+use WC_Product;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Class ChannelVisibilityBlock
+ *
+ * @package Automattic\WooCommerce\GoogleListingsAndAds\Admin\Product
+ */
+class ChannelVisibilityBlock {
+
+	public const PROPERTY = 'google_listings_and_ads__channel_visibility';
+
+	/**
+	 * @var ProductHelper
+	 */
+	protected $product_helper;
+
+	/**
+	 * ChannelVisibilityBlock constructor.
+	 *
+	 * @param ProductHelper $product_helper
+	 */
+	public function __construct( ProductHelper $product_helper ) {
+		$this->product_helper = $product_helper;
+	}
+
+	/**
+	 * Register hooks for querying and updating product via REST APIs.
+	 */
+	public function register_hooks(): void {
+		// https://github.com/woocommerce/woocommerce/blob/8.5.0/plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-products-v2-controller.php#L182-L192
+		add_filter( 'woocommerce_rest_prepare_product_object', [ $this, 'prepare_data' ], 10, 2 );
+
+		// https://github.com/woocommerce/woocommerce/blob/8.5.0/plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-crud-controller.php#L200-L207
+		// https://github.com/woocommerce/woocommerce/blob/8.5.0/plugins/woocommerce/includes/rest-api/Controllers/Version3/class-wc-rest-crud-controller.php#L247-L254
+		add_filter( 'woocommerce_rest_insert_product_object', [ $this, 'update_data' ], 10, 2 );
+	}
+
+	/**
+	 * Get channel visibility data from the given product and add it to the given response.
+	 *
+	 * @param Response   $response Response to be added channel visibility data.
+	 * @param WC_Product $product WooCommerce product to get data.
+	 *
+	 * @return Response
+	 */
+	public function prepare_data( Response $response, WC_Product $product ): Response {
```

Comment:
The key call stacks should show the `$product` is always an instance of `WC_Product`:

- https://github.com/woocommerce/woocommerce/blob/8.5.0/plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-products-v2-controller.php#L149-L151
- https://github.com/woocommerce/woocommerce/blob/8.5.0/plugins/woocommerce/includes/wc-product-functions.php#L64-L74

Also, the method of `$product->is_visible()` call is implemented in [`WC_Product`](https://github.com/woocommerce/woocommerce/blob/8.5.0/plugins/woocommerce/includes/abstracts/abstract-wc-product.php#L1584), but not in the `WC_Data`. I think it would be clearer than using the abstract class `WC_Data ` to explicitly indicate that this method expects the incoming `$product` to be an instance of `WC_Product`, and also ensures that subsequent calls are valid.

Reply:
Yes, from what I can tell, it seems like it will always be `WC_Product`. However, when I check examples in other plugins calling this filter, they also expect `WC_Data`.

See here: 

- https://github.com/woocommerce/woocommerce-brands/blob/develop/includes/class-wc-brands.php#L845
- https://github.com/woocommerce/woocommerce-brands/blob/f82f270efb374b746883a3b30374583c5fb9d9c5/includes/class-wc-brands.php#L873

Reply:
Thanks for the suggestion mentioned in https://github.com/woocommerce/google-listings-and-ads/pull/2214#pullrequestreview-1842430278, adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/2214/commits/34abdb217c324689c8c537402ef1e2acc09e46db.